tsl/src/continuous_aggs/refresh.c

```diff
@@ -651,29 +658,26 @@ continuous_agg_refresh_internal(const ContinuousAgg *cagg,
 	if (refresh_window.start >= refresh_window.end)
 	{
 		emit_up_to_date_notice(cagg, callctx);
+
+		if ((rc = SPI_finish()) != SPI_OK_FINISH)
+			elog(ERROR, "SPI_finish failed: %s", SPI_result_code_string(rc));
+
 		return;
 	}
 
 	/* Process invalidations in the hypertable invalidation log */
 	invalidation_process_hypertable_log(cagg, refresh_window.type);
 
-	/* Start a new transaction. Note that this invalidates previous memory
-	 * allocations (and locks). */
-	PopActiveSnapshot();
-	CommitTransactionCommand();
-	/* See PG commit:84f5c2908dad81e8622b0406beea580e40bb03ac
-	 * When we manage multiple transactions in a procedure, ensure that
-	 * an active outer snapshot exists prior to executing SPI
-	 * (see EnsurePortalSnapshotExists)
-	 */
-	StartTransactionCommand();
-	PushActiveSnapshot(GetTransactionSnapshot());
+	/* Commit and Start a new transaction */
+	SPI_commit_and_chain();
 
 	cagg = ts_continuous_agg_find_by_mat_hypertable_id(mat_id);
 
 	if (!process_cagg_invalidations_and_refresh(cagg, &refresh_window, callctx, INVALID_CHUNK_ID))
 		emit_up_to_date_notice(cagg, callctx);
-	PopActiveSnapshot();
+
```

Comment:
With this change, I think we can remove the SPI_connect and SPI_finish in materialize.c code. I don't see the reason to do it there. Is there any?

Reply:
Is better to leave as is because the `continuous_agg_update_materialization` is also invoked (in sub-calls) by the `continuous_agg_refresh_chunk` that leaves in the `timescaledb_experimental` schema and doesn't have the SPI_connect there